src/mococrw/symmetric_crypto.h

```diff
@@ -0,0 +1,367 @@
+     *
+     * @param padding type of padding.
+     * @return builder instance
+     */
+    AESCipherBuilder &setPadding(SymmetricCipherPadding padding);
+
+    /**
+     * Set IV
+     *
+     * When creating a cipher, this method must be used to provide the IV. If cipher is created for
+     * decryption, call of this method is mandatory. If cipher is created for encryption, random IV
+     * of maximum supported length (128 bit) is generated by the builder by default. user still can
+     * override generated by default IV with his own value.
+     *
+     * @note If you set custom IV for encryption, make sure that it is supported by the type of
+     * cipher you are building. Also, use cryptographically secure RNG utility::cryptoRandomBytes().
+     *
+     * @param iv the initialization vector. Usually, you get this value together with the ciphertext and password.
+     * @return
+     */
+    AESCipherBuilder &setIV(const std::vector<uint8_t> &iv);
+
+    /**
+     * Set length of authentication tag for authenticated encryption.
+     *
+     * Use this method to change default length of authentication tag which encryptor will return
+     * in getAuthTag(). Default value is 128 bits.
+     *
+     * @note Do not use when creating a cipher for decryption. It does not make sense and will throw.
+     *
+     * @warning Please note that security of authenticated encryption directly depends on length of
+     * authentication tag. If you think that you have valid reasons for using tag length less than
+     * 64bit, please consult with Appendix C https://csrc.nist.gov/publications/detail/sp/800-38d/final.
+     *
+     * @param length
+     * @return
+     */
+    AESCipherBuilder &setAuthTagLength(size_t length);
+
+    /**
+     * Create cipher for encryption.
+     *
+     * @return encryptor
+     */
+    std::shared_ptr<AESCipher> buildEncryptor();
+
+    /**
+     * Create cipher for decryption.
+     *
+     * @return dencryptor
+     */
+    std::shared_ptr<AESCipher> buildDecryptor();
+
+    /**
+     * Create authenticated cipher for encryption.
+     *
+     * This method must be called to build authenticated cipher such as SymmetricCipherMode::GCM.
+     *
+     * @return encryptor
+     */
+    std::shared_ptr<AuthenticatedAESCipher> buildAuthenticatedEncryptor();
+
+    /**
+     * Create authenticated cipher for decryption.
+     *
+     * This method must be called to build authenticated cipher such as SymmetricCipherMode::GCM.
+     *
+     * @return dencryptor
+     */
+    std::shared_ptr<AuthenticatedAESCipher> buildAuthenticatedDecryptor();
+
+private:
+    // IV length differs from mode to mode but 16 bytes is the maximum currently supported by OpenSSL
+    static const size_t _EVP_MAX_IV_LENGTH = 16;
+    static const size_t DefaultAuthTagLength = 16;
+
+    std::vector<uint8_t> _iv = utility::cryptoRandomBytes(_EVP_MAX_IV_LENGTH);
```

Comment:
Good point. Will fix.

Reply:
Your fix for this didn't work, because you still set `_iv` in the builder object when `_iv.size() == 0`, so the second call to `buildEncryptor()` would still have re-used the same IV.